cns/restserver/api.go

```diff
@@ -763,6 +765,71 @@ func (service *HTTPRestService) setOrchestratorType(w http.ResponseWriter, r *ht
 	logger.Response(service.Name, resp, resp.ReturnCode, err)
 }
 
+// getHomeAz retrieves home AZ of host
+func (service *HTTPRestService) getHomeAz(w http.ResponseWriter, r *http.Request) {
+	logger.Printf("[Azure CNS] getHomeAz")
+	logger.Request(service.Name, "getHomeAz", nil)
+	ctx := r.Context()
+
+	switch r.Method {
+	case http.MethodGet:
+		homeAzResponse, err := service.nma.GetHomeAz(ctx)
+		if err != nil {
+			apiError := nma.Error{}
+			if ok := errors.As(err, &apiError); ok {
+				switch apiError.StatusCode() {
+				case http.StatusNotImplemented:
+					returnMessage := fmt.Sprintf("[Azure CNS] Error. NMAgent does not support %s api.", nma.GetHomeAzAPIName)
+					returnCode := types.NmAgentUnSupportedAPIError
+					service.setResponse(w, returnCode, cns.GetHomeAzResponse{
+						Response: cns.Response{ReturnCode: returnCode, Message: returnMessage},
+					})
+					return
+				case http.StatusInternalServerError:
+					returnMessage := "[Azure CNS] Error. GetHomeAz failed due to Nmagent server internal error."
+					returnCode := types.NmAgentInternalServerError
+					service.setResponse(w, returnCode, cns.GetHomeAzResponse{
+						Response: cns.Response{ReturnCode: returnCode, Message: returnMessage},
+					})
+					return
+
+				case http.StatusUnauthorized:
+					returnMessage := "[Azure CNS] Error. GetHomeAz failed to authenticate with OwningServiceInstanceId"
+					returnCode := types.StatusUnauthorized
+					service.setResponse(w, returnCode, cns.GetHomeAzResponse{
+						Response: cns.Response{ReturnCode: returnCode, Message: returnMessage},
+					})
+					return
+
+				default:
+					returnMessage := fmt.Sprintf("[Azure CNS] Error. GetHomeAz failed with StatusCode: %d", apiError.StatusCode())
+					returnCode := types.UnexpectedError
+					service.setResponse(w, returnCode, cns.GetHomeAzResponse{
+						Response: cns.Response{ReturnCode: returnCode, Message: returnMessage},
+					})
+					return
+				}
+			}
+			returnCode := types.UnexpectedError
+			returnMessage := fmt.Sprintf("[Azure CNS] Error. getHomeAz failed %v", err)
+			service.setResponse(w, returnCode, cns.GetHomeAzResponse{
+				Response: cns.Response{ReturnCode: returnCode, Message: returnMessage},
+			})
+			return
+		}
```

Comment:
I really don't want us to be in the business of re-litigating every error we get from something and figuring out what it means. That just spreads intrinsic knowledge about how subsystems work out to components that shouldn't have it (creating N places to change when that knowledge changes instead of 1). My intention with the nmagent client was that the error tells you what went wrong, so ideally the error handling looks as boring as:

```go
homeAzResponse, err := service.nma.GetHomeAz(ctx)
if err != nil {
       service.setResponse(w, returnCode, cns.GetHomeAzResponse{
              Response: cns.Response{
                     ReturnCode: NMAgentFailed, // constantize this somewhere
                     Message: err.Error()
              }
       })
       return
}
```

This isolates things to what we know in this handler: we tried to do something with NMAgent, but it failed. We don't know what, but the error has all the detail you would need.

The current approach also has too much expectation that the error returned is an `nmagent.Error`. Other types can (and will) be returned.

Why do we need error codes for each possible outcome?